File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.py
class Solution:
    def findAllRecipes(self, recipes: List[str], ingredients: List[List[str]], supplies: List[str]) -> List[str]:
        graph = defaultdict(list)
        indegree = defaultdict(int)
        q = deque()
        result = []
        
        # everything in supplies has an indegree of 0
        #reciepe is the end node , supply is start
        for s in supplies:
            indegree[s] = 0
            q.append(s)
            
        for i, r in enumerate(recipes):
            for ing in ingredients[i]:
                graph[ing].append(r)
                indegree[r] += 1
                
        while q:
            node = q.pop()
            if node in recipes:
                result.append(node)
            for n in graph[node]:
                indegree[n] -= 1
                if indegree[n] == 0:
                    q.append(n)
                    
        return result

File: 2115-find-all-possible-recipes-from-given-supplies/2115-find-all-possible-recipes-from-given-supplies.js
var findAllRecipes = function(recipes, ingredients, supplies) {
    const availableSupplies = new Set(supplies);
    const dependency_Ing_ToReceipe = new Map();
    const indegree = new Array(recipes.length).fill(0);

    for (let i = 0; i < recipes.length; i++) {
        for (const ing of ingredients[i]) {
            if (!availableSupplies.has(ing)) {
                if (!dependency_Ing_ToReceipe.has(ing)) dependency_Ing_ToReceipe.set(ing, []);
                dependency_Ing_ToReceipe.get(ing).push([recipes[i],i]);
                indegree[i]++;
            }
        }
    }

    const queue = recipes.filter((_, i) => indegree[i] === 0);
    const result = [];
    while (queue.length > 0) {
        const currentRecipe = queue.shift();
        result.push(currentRecipe);
        const deps = dependency_Ing_ToReceipe.get(currentRecipe) || [];
        for (const dep of deps) {
            const depIdx = dep[1]
            indegree[depIdx]--;
            if (indegree[depIdx] === 0) queue.push(dep[0]);
        }
    }
    
    return result;
};
